include/net/lwm2m_client_utils.h

```diff
@@ -144,37 +144,115 @@ int lwm2m_init_location(void);
+
+/** Firmware update callback events. */
+enum lwm2m_fota_event_id {
+	/** Download process started */
+	LWM2M_FOTA_DOWNLOAD_START,
+	/** Download process finished */
+	LWM2M_FOTA_DOWNLOAD_FINISHED,
+	/** Request for update new image */
+	LWM2M_FOTA_UPDATE_IMAGE_REQ,
+	/** Fota process fail or cancelled  */
+	LWM2M_FOTA_UPDATE_ERROR
+};
+
+/** Event data for id LWM2M_FOTA_DOWNLOAD_START. */
+struct lwm2m_fota_download_start {
+	/** Object Instance  id for event */
+	uint16_t obj_inst_id;
+};
+
+/** Event data for id LWM2M_FOTA_DOWNLOAD_FINISHED. */
+struct lwm2m_fota_download_finished {
+	/** Object Instance  id for event */
+	uint16_t obj_inst_id;
+	/** DFU type */
+	int dfu_type;
+};
+
+/** Event data for id LWM2M_FOTA_UPDATE_IMAGE_REQ. */
+struct lwm2m_fota_update_request {
+	/** Object Instance  id for event */
+	uint16_t obj_inst_id;
+	/** DFU type */
+	int dfu_type;
+};
+
+/** Event data for id LWM2M_FOTA_UPDATE_ERROR. */
+struct lwm2m_fota_update_failure {
+	/** Object Instance  id for event */
+	uint16_t obj_inst_id;
+	/** FOTA failure result */
+	uint8_t update_failure;
+};
+
+struct lwm2m_fota_event {
+	/** Fota event id and indicate used Data structure */
+	enum lwm2m_fota_event_id id;
+	union {
+		/** LWM2M_FOTA_DOWNLOAD_START */
+		struct lwm2m_fota_download_start download_start;
+		/** LWM2M_FOTA_DOWNLOAD_FINISHED */
+		struct lwm2m_fota_download_finished download_ready;
+		/** LWM2M_FOTA_UPDATE_IMAGE_REQ */
+		struct lwm2m_fota_update_request update_req;
+		/** LWM2M_FOTA_UPDATE_ERROR */
+		struct lwm2m_fota_update_failure failure;
+	};
+};
+
 /**
- * @brief Firmware update state change event callback.
+ * @brief Firmware update event callback.
  *
- * @param[in] update_state LwM2M Firmware Update object states
+ * @param[in] event LwM2M Firmware Update object event structure
  *
- * @return Callback returns a negative error code (errno.h) indicating
- *         reason of failure or 0 for success.
- */
-typedef int (*lwm2m_firmware_get_update_state_cb_t)(uint8_t update_state);
-
-/**
- * @brief Set event callback for firmware update changes.
+ * Callback is used for following firmware update states and prepare app ready for update.
  *
- * LwM2M clients use this function to register a callback for receiving the
- * update state changes when performing a firmware update.
+ * Event handler is getting event callback when Firmware update utils library change states.
  *
- * @param[in] cb A callback function to receive firmware update state changes or NULL for disable.
+ * LWM2M_FOTA_DOWNLOAD_START: Indicate that download or Upload new image is started.
+ *
+ * LWM2M_FOTA_DOWNLOAD_FINISHED: Indicate that Image is delivered.
+ *
+ * LWM2M_FOTA_UPDATE_IMAGE_REQ: Update resource is excuted and utils libray want update binary
+ * Callback handler return 0 for accept update process. Otherwise return any negative for cancel or
+ * positive value postpone new request in seconds.
```

Comment:
```suggestion
 * Callback handler may return zero to grant permission for update. Otherwise it may return negative error code to cancel the update or
 * positive value to indicate that update should be postponed that amount of seconds.
```